### PR TITLE
RELATED: BB-2211 Format preset dropdown

### PR DESCRIPTION
**Pull request by strouhaljan**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [ ] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [ ] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [ ] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [ ] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [ ] Successful `extended test - examples`
- [ ] Successful `extended test - storybook`
- [ ] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-analytical-designer:
- gdc-dashboards:

# Related Jira tasks

### CODE REVIEW

File: src/translations/en-US.json [hunk 1]
```diff
@@ -856,5 +856,55 @@
         "value": "If the filter condition includes 0, blank values will be treated as 0 and included in calculations.",
         "comment": "Checkbox label tooltip. 'blank values' are data rows without any value (empty), which can be considered to be 0 during calculations. If the filter is set to '< 1' for example, it includes values of 0 (less than one) and therefore all the blank (empty) values are going to be considered to be 0 in calculations.",
         "limit": 0
+    },
+    "measureNumberFormat.inherit.optionLabel": {
```
Comment (strouhaljan) [inline src/translations/en-US.json:860]: I believe this is not used anywhere anymore (inherit preset is not hardcoded anymore).

File: stories/internal/MeasureNumberFormat.tsx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+import { IFormatPreset, IToggleButtonProps } from "../../src/components/menu/measureNumberFormat/typings";
+import { screenshotTarget } from "../utils/wrap";
+import { MeasureNumberFormat } from "../../src/components/menu/measureNumberFormat/MeasureNumberFormat";
+import { IFormatTemplate } from "../../src/components/menu/measureNumberFormat/CustomFormatTemplates";
+
+storiesOf("Internal/Measure Number Format", module).add("Measure Number Format", () => {
+    const documentationLink =
+        "https://help.gooddata.com/doc/en/reporting-and-dashboards/reports/working-with-reports/formatting-numbers-in-reports";
+    const presets: IFormatPreset[] = [
```
Comment (strouhaljan) [inline stories/internal/MeasureNumberFormat.tsx:16]: Add empty line.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7282/ : FAILURE in 4m 26s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3284/ : SUCCESS in 3s

File: src/components/menu/measureNumberFormat/CustomFormatPreview.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
+// (C) 2020 GoodData Corporation
+import * as React from "react";
+import { WrappedComponentProps } from "react-intl";
+import InputWithNumberFormat from "@gooddata/goodstrap/lib/Form/InputWithNumberFormat";
+
+import { CustomFormatExtendedPreview } from "./CustomFormatExtendedPreview";
+import { FormattedPreview } from "./FormattedPreview";
+import * as numberJS from "@gooddata/numberjs";
```
Comment (strouhaljan) [inline src/components/menu/measureNumberFormat/CustomFormatPreview.tsx:8]: Move this line up among other 3rd party libraries imports.

File: src/components/menu/measureNumberFormat/CustomFormatPreviewRows.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
+// (C) 2020 GoodData Corporation
+
```
Comment (strouhaljan) [inline src/components/menu/measureNumberFormat/CustomFormatPreviewRows.tsx:2]: Remove empty line.

File: src/components/menu/measureNumberFormat/CustomFormatPreviewRows.tsx [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import * as numberJS from "@gooddata/numberjs";
```
Comment (strouhaljan) [inline src/components/menu/measureNumberFormat/CustomFormatPreviewRows.tsx:3]: Move gooddata library import below 3rd party library import (react).

File: src/components/menu/measureNumberFormat/CustomFormatTemplates.tsx [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+interface ITemplateDropdownItemState {
+    displayHelp: boolean;
+}
+
+function templateDropdownItemId(template: IFormatTemplate): string {
+    return `gd-format-preset-template-${template.localIdentifier}`;
+}
+
+class TemplateDropdownItem extends React.Component<ITemplateDropdownItemProps, ITemplateDropdownItemState> {
+    public state = {
```
Comment (strouhaljan) [inline src/components/menu/measureNumberFormat/CustomFormatTemplates.tsx:45]: Isn't the state supposed to be private?

File: src/components/menu/measureNumberFormat/MeasureNumberFormat.tsx [hunk 7]
```diff
@@ -0,0 +1,209 @@
+// (C) 2020 GoodData Corporation
+import * as React from "react";
+import { ISeparators } from "@gooddata/numberjs";
+import { injectIntl, IntlShape, WrappedComponentProps } from "react-intl";
+
+import { IFormatPreset, IToggleButtonProps, PresetType } from "./typings";
+import { MeasureNumberFormatDropdown } from "./MeasureNumberFormatDropdown";
+
```
Comment (strouhaljan) [inline src/components/menu/measureNumberFormat/MeasureNumberFormat.tsx:8]: Remove empty line.

File: src/components/menu/measureNumberFormat/MeasureNumberFormat.tsx [hunk 8]
```diff
@@ -0,0 +1,209 @@
+// (C) 2020 GoodData Corporation
+import * as React from "react";
+import { ISeparators } from "@gooddata/numberjs";
```
Comment (strouhaljan) [inline src/components/menu/measureNumberFormat/MeasureNumberFormat.tsx:3]: Move gooddata library import below 3rd party library import (react-intl).

File: src/components/menu/measureNumberFormat/MeasureNumberFormatDropdown.tsx [hunk 9]
```diff
@@ -0,0 +1,67 @@
+// (C) 2020 GoodData Corporation
+import * as React from "react";
+import Overlay from "@gooddata/goodstrap/lib/core/Overlay";
+import * as numberJS from "@gooddata/numberjs";
+import { injectIntl, IntlShape } from "react-intl";
```
Comment (strouhaljan) [inline src/components/menu/measureNumberFormat/MeasureNumberFormatDropdown.tsx:5]: Fix order of imports.

File: src/index.ts [hunk 10]
```diff
@@ -164,4 +167,7 @@ export {
     HeaderPredicateFactory,
     MappingHeader,
     withJsxExport,
+    MeasureNumberFormat,
+    MeasureNumberFormatType,
+    IPositioning,
```
Comment (tomassatinsky) [inline src/index.ts:172]: let's export also `IFormatPreset` so we can use it in AD

File: stories/utils/wrap.tsx [hunk 11]
```diff
@@ -26,3 +26,11 @@ export function wrap(
         </IntlWrapper>
     );
 }
+
+export function screenshotTarget(component: JSX.Element, width = 400, height = 400) {
```
Comment (tomassatinsky) [inline stories/utils/wrap.tsx:30]: what is this? Is it re-usable?

File: styles/scss/measureNumberFormat.scss [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+        & > .gd-format-preset-preview {
+            color: $gd-color-text-dimmed;
+            font-weight: normal;
+            margin-right: 5px;
+            text-align: right;
+        }
+    }
+}
+
+$dialog_width: 250px;
```
Comment (tomassatinsky) [inline styles/scss/measureNumberFormat.scss:25]: is it possible to these variables to pollute global scope and possibly conflict with the same name elsewhere? A more specific name could prevent it.

File: src/translations/en-US.json [hunk 13]
```diff
@@ -856,5 +856,55 @@
...
+        "limit": 0
+    },
+    "measureNumberCustomFormatDialog.howToFormat": {
+        "value": "How to format?",
+        "comment": "Link to help",
+        "limit": 0
+    },
+    "measureNumberCustomFormatDialog.definition": {
+        "value": "Definition",
+        "comment": "Title above the textarea for measure format definition",
```
Comment (tomassatinsky) [inline src/translations/en-US.json:902]: I'm afraid "above the textarea" would not help the translators who don't have the picture. Something "Header of format definition field" would be sufficient.

File: src/translations/en-US.json [hunk 14]
```diff
@@ -856,5 +856,55 @@
...
+        "limit": 0
+    },
+    "measureNumberCustomFormatDialog.title": {
+        "value": "Custom format",
+        "comment": "The title of the dialog for setting a custom measure format",
+        "limit": 0
+    },
+    "measureNumberCustomFormatDialog.template.preview.title": {
+        "value": "Format preview",
+        "comment": "The title of the help bubble for custom measure format template",
```
Comment (tomassatinsky) [inline src/translations/en-US.json:882]: I guess the translators will have no clue what "custom measure format template" is. Let's try something generic. For instance "Header of a section showing format previews "

Comment (tomassatinsky) [reply src/translations/en-US.json:882]: Please check all the comments with this in mind

File: src/components/menu/measureNumberFormat/CustomFormatDialog.tsx [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+                                disabled={this.isApplyButtonDisabled()}
+                            />
+                        </div>
+                    </div>
+                </div>
+            </Overlay>
+        );
+    }
+
+    private renderFormatInput = () => (
```
[the remark this answers is not in the document]
Comment (yanlipnican) [reply src/components/menu/measureNumberFormat/CustomFormatDialog.tsx:92]: I think way to go is splitting to components which are private in same file. Or we can create barrel file which exports only top level component and have other components in separate files

File: stories/internal/MeasureNumberFormat.tsx [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+import { action } from "@storybook/addon-actions";
+import { ISeparators } from "@gooddata/numberjs";
+import { screenshotWrap } from "@gooddata/test-storybook";
+
+import { IFormatPreset, IToggleButtonProps } from "../../src/components/menu/measureNumberFormat/typings";
+import { screenshotTarget } from "../utils/wrap";
+import { MeasureNumberFormat } from "../../src/components/menu/measureNumberFormat/MeasureNumberFormat";
+import { IFormatTemplate } from "../../src/components/menu/measureNumberFormat/CustomFormatTemplates";
+
+storiesOf("Internal/Measure Number Format", module).add("Measure Number Format", () => {
```
Comment (tomassatinsky) [inline stories/internal/MeasureNumberFormat.tsx:13]: The component is in the "menu" dir in the source code so I guess it should be in the "menu" dir in storybook too.

File: src/components/menu/measureNumberFormat/CustomFormatDialog.tsx [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+                        <div className="gd-measure-custom-format-dialog-header">
+                            <span>{intl.formatMessage({ id: "measureNumberCustomFormatDialog.title" })}</span>
+                        </div>
+                        <div className="gd-measure-custom-format-dialog-content">
+                            {this.renderFormatInput()}
+                            {documentationLink && this.renderHelp()}
+                            <CustomFormatPreview
+                                formatString={formatString}
+                                separators={separators}
+                                intl={intl}
```
Comment (yanlipnican) [inline src/components/menu/measureNumberFormat/CustomFormatDialog.tsx:70]: Don't pass inlt by prop, rather use inject inlt HOC for component

File: stories/utils/wrap.tsx (hunk 11 above)
Comment (xMort) [reply stories/utils/wrap.tsx:30]: Yep, plan is to have this reusable, as many stories repeat this code many times inside. I wanted to use this in place of the repeated code but it is probably not good idea to pollute this pull request with it.

File: stories/internal/MeasureNumberFormat.tsx (hunk 16 above)
Comment (xMort) [reply stories/internal/MeasureNumberFormat.tsx:13]: Possibly we could find better directory for it as I'm not sure "menu" is that great. Originally I had menu in name of the components but removed it. I wanted to move it in different directory after mutual discussion.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7291/ : FAILURE in 1m 38s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3297/ : SUCCESS in 3s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7292/ : SUCCESS in 5m 34s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3298/ : SUCCESS in 3s

File: styles/scss/measureNumberFormat.scss (hunk 12 above)
Comment (tomassatinsky) [reply styles/scss/measureNumberFormat.scss]: ...and the same for the one below

File: src/index.ts (hunk 10 above)
Comment (strouhaljan) [reply src/index.ts:172]: `IFormatPreset` is already exported within `MeasureNumberFormatType`. The question is, whether this is not a weird way.

File: src/components/menu/measureNumberFormat/CustomFormatDialog.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (strouhaljan) [reply src/components/menu/measureNumberFormat/CustomFormatDialog.tsx]: I did some splitting, a bit experimental, more than we usually do :)

File: src/components/menu/measureNumberFormat/MeasureNumberFormat.tsx [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+
+        this.presets = buildPresets(props);
+        this.state = {
+            selectedPreset: this.findSelectedPreset(),
+            showDropdown: false,
+            showCustomFormatDialog: false,
+        };
+    }
+
+    public UNSAFE_componentWillReceiveProps(nextProps: MeasureNumberFormatProps) {
```
[the remark this answers is not in the document]
Comment (strouhaljan) [reply src/components/menu/measureNumberFormat/MeasureNumberFormat.tsx]: Avoided.

File: src/components/menu/measureNumberFormat/MeasureNumberFormat.tsx [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+        return (
+            <>
+                <ToggleButton
+                    text={buttonText}
+                    isOpened={showDropdown || showCustomFormatDialog}
+                    toggleDropdown={this.toggleDropdownOpened}
+                />
+                {showDropdown && (
+                    <PresetsDropdown
+                        presets={[...presets, this.getCustomFormatPreset()]}
```
Comment (xMort) [inline src/components/menu/measureNumberFormat/MeasureNumberFormat.tsx:83]: This producers new array each time functional component is called. We decided to not do memoization if we don't do any measurement of performance to have more readable code?

Comment (strouhaljan) [reply src/components/menu/measureNumberFormat/MeasureNumberFormat.tsx:83]: Changed the way presets are rendered.

File: stories/internal/MeasureNumberFormat.tsx (hunk 16 above)
Comment (strouhaljan) [reply stories/internal/MeasureNumberFormat.tsx:13]: I was unable to come up with a proper location so I've created it's own top-level directory instead.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7301/ : SUCCESS in 5m 50s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3310/ : SUCCESS in 3s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7302/ : SUCCESS in 2m 46s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3311/ : SUCCESS in 3s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by tomassatinsky** [comment]
LGTM

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3831/ : UNSTABLE in 20m 17s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-fossa-scanning-zuul-el7/10/ : SUCCESS in 39s (non-voting)
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7305/ : SUCCESS in 11m 10s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1785/ : SUCCESS in 25s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/2622/ : SUCCESS in 15m 59s

**Comment by strouhaljan** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3832/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3317/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3196/ : UNSTABLE in 11m 38s

**Comment by strouhaljan** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7312/ : SUCCESS in 4m 08s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3322/ : SUCCESS in 3s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3834/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3323/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3198/ : UNSTABLE in 16m 33s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by strouhaljan** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7322/ : FAILURE in 9m 48s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3334/ : SUCCESS in 26s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3841/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3335/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3203/ : UNSTABLE in 20m 13s

**Comment by strouhaljan** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7327/ : SUCCESS in 10m 51s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3340/ : SUCCESS in 3s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3843/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3341/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3205/ : SUCCESS in 12m 58s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3844/ : SUCCESS in 18m 51s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-fossa-scanning-zuul-el7/15/ : SUCCESS in 19s (non-voting)
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7329/ : SUCCESS in 3m 49s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1790/ : SUCCESS in 13s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/2629/ : SUCCESS in 18m 11s

**Pull request merged**